scripts/benchmark/trainer-benchmark.py

```diff
@@ -0,0 +1,435 @@
+# --target-metric-key train_samples_per_second --repeat-times 1 --variations \
+# '|--fp16|--bf16' '--tf32 0|--tf32 1' --report-metric-keys train_loss \
+# --repeat-times 1 --base-variation '--tf32 0'
+#
+# and here is a possible output:
+#
+#
+# | Variation       |     Train |   Diff |   Train |
+# |                 |   samples |      % |    loss |
+# |                 |       per |        |         |
+# |                 |    second |        |         |
+# |:----------------|----------:|-------:|--------:|
+# | --tf32 0        |    285.11 |      0 |    2.51 |
+# | --tf32 1        |    342.09 |     20 |    2.51 |
+# | --fp16 --tf32 0 |    423.49 |     49 |    2.51 |
+# | --fp16 --tf32 1 |    423.13 |     48 |    2.51 |
+# | --bf16 --tf32 0 |    416.80 |     46 |    2.52 |
+# | --bf16 --tf32 1 |    415.87 |     46 |    2.52 |
+#
+#
+# So you can quickly compare the different outcomes.
+#
+# Typically running each experiment once is enough, but if the environment is unstable you can
+# re-run each multiple times, e.g., 3 using --repeat-times 3 and it will report the averaged results.
+#
+# By default it'll use the lowest result as the base line to use as 100% and then compare the rest to
+# it as can be seen from the table above, but you can also specify which combination is the one to use as
+# the baseline, e.g., to change to another entry use: --base-variation '--tf32 1 --fp16 0'
+#
+# --target-metric-key is there to tell the program which metrics to compare - the different metric keys are
+# inside output_dir/all_results.json. e.g., to measure eval performance instead of train use:
+#    --target-metric-key eval_samples_per_second
+# but of course you will need to adjust the --base-cmd value in the example to perform evaluation as
+# well (as currently it doesn't)
+#
+
+import argparse
+import datetime
+import io
+import itertools
+import json
+import math
+import os
+import platform
+import re
+import shlex
+import subprocess
+import sys
+from pathlib import Path
+from statistics import fmean
+
+import pandas as pd
+import torch
+from tqdm import tqdm
+
+import transformers
+
+
+nan = float("nan")
+
+
+class Tee:
+    """
+    A helper class to tee print's output into a file.
+    Usage:
+    sys.stdout = Tee(filename)
+    """
+
+    def __init__(self, filename):
+        self.stdout = sys.stdout
+        self.file = open(filename, "a")
+
+    def __getattr__(self, attr):
+        return getattr(self.stdout, attr)
+
+    def write(self, msg):
+        self.stdout.write(msg)
+        # strip tqdm codes
+        self.file.write(re.sub(r"^.*\r", "", msg, 0, re.M))
+
+
+def get_orig_cmd(max_width=80, full_python_path=False):
+    """
+    Return the original command line string that can be replayed nicely and wrapped for 80 char width
+    Args:
+        - max_width: the width to wrap for. defaults to 80
+        - full_python_path: whether to replicate the full path or just the last part (i.e. `python`). default to `False`
```

Comment:
I don't understand the point of writing docstrings if they're not following the format we use.

Reply:
that's a very odd thing to say. The purpose of docstrings is to let users know how to use the function. The exact format is completely different story.

As I had no idea whether this tool will be found useful and accepted I focused on functionality first and were planning to worry about conformance later.

Reply:
Why have inconsistent docstring format is what I was trying to say? Why spend the time writing a docstring that does not fit with the rest of the library?
If you are in a hurry and waiting to see if the feature is accepted, I would have understood having no doscting at all, it's just that since you went to the trouble of writing one, I'd expect it to fit the rest of the lib.

Reply:
I understand, then please just say - this needs to be reformatted to our standard to be accepted.

But saying that something is pointless because it's not up to some random standard is hard to digest.

I will reformat this to the HF standard now that I know that this tool is wanted.

Reply:
Sorry if my initial comment felt rude, that was not my intention. Please do reformat the docstring to the usual format :-)